kernel/init.c

```diff
@@ -280,36 +315,41 @@ static void z_sys_init_run_level(enum init_level level)
 		const struct device *dev = entry->dev;
 
 		if (dev != NULL) {
-			int rc = 0;
-
-			if (entry->init_fn.dev != NULL) {
-				rc = entry->init_fn.dev(dev);
-				/* Mark device initialized. If initialization
-				 * failed, record the error condition.
-				 */
-				if (rc != 0) {
-					if (rc < 0) {
-						rc = -rc;
-					}
-					if (rc > UINT8_MAX) {
-						rc = UINT8_MAX;
-					}
-					dev->state->init_res = rc;
-				}
-			}
-
-			dev->state->initialized = true;
-
-			if (rc == 0) {
-				/* Run automatic device runtime enablement */
-				(void)pm_device_runtime_auto_enable(dev);
-			}
+			do_device_init(entry);
 		} else {
 			(void)entry->init_fn.sys();
 		}
 	}
 }
 
+
+int z_impl_device_init(const struct device *dev)
+{
+	const struct init_entry *entry;
+
+	if (dev == NULL) {
+		return -ENOENT;
+	}
+
+	for (entry = __deferred_init_start; entry <= __deferred_init_end; entry++) {
+		if (entry->dev == dev) {
+			return do_device_init(entry);
+		}
+	}
```

Comment:
please use iterable sections

Reply:
Yes - I can't remember the exact macro off the top of my head `ITERABLE_SECTION_FOR_EACH()` maybe? There is likely an ALT version that should allow for pointer aliasing if necessary